src/include/txn_inline.h

```diff
@@ -631,6 +631,57 @@ __wt_txn_oldest_id(WT_SESSION_IMPL *session)
     }
 }
 
+/*
+ * __wt_txn_pinned_stable_timestamp --
+ *     Get the first timestamp that can be written to the disk for precise checkpoint.
+ */
+static WT_INLINE void
+__wt_txn_pinned_stable_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *pinned_stable_tsp)
+{
+    WT_TXN_GLOBAL *txn_global;
+    wt_timestamp_t checkpoint_ts, pinned_stable_ts;
+    bool has_stable_timestamp;
+
+    txn_global = &S2C(session)->txn_global;
+
+    /*
+     * There is no need to go further if no stable timestamp has been set yet.
+     */
+    WT_ACQUIRE_READ_WITH_BARRIER(has_stable_timestamp, txn_global->has_stable_timestamp);
+    if (!has_stable_timestamp) {
+        *pinned_stable_tsp = WT_TS_NONE;
+        return;
+    }
+
+    /*
+     * It is important to ensure we only read the global stable timestamp once. Otherwise, we may
+     * return a stable timestamp that is larger than the checkpoint timestamp. For example, the
+     * first time we read the global stable timestamp as 100 and set it to the local variable
+     * disaggregated_stable_ts. If the checkpoint timestamp is 110 and the second time we read the
+     * global stable timestamp as 120, we will return 120 instead of the checkpoint timestamp 110.
+     */
+    WT_ACQUIRE_READ_WITH_BARRIER(pinned_stable_ts, txn_global->stable_timestamp);
+
+    if (!__wt_conn_is_disagg(session)) {
+        *pinned_stable_tsp = pinned_stable_ts;
+        return;
+    }
+
+    /*
+     * The read of checkpoint timestamp needs to be carefully ordered: it needs to be after we have
```

Comment:
Copy-paste bug, I think this needs to reference the stable timestamp now (rather than pinned)?